Add wetland scheme of Zhang et al 2022 WRR

cenlinhe opened the pull request:
This PR is a refined version of the code implementation for the wetland scheme from Zhang et al. 2022 (https://doi.org/10.1029/2021WR030573) following an earlier PR (https://github.com/NCAR/noahmp/pull/97). This PR resolves the code merging conflicts in the earlier PR.

Original PR and implementation: Zhe Zhang (NCAR)
Refined PR and implementation: Cenlin He (NCAR)

The wetland code has been refactored to the newest version of Noah-MP LSM.
To activate the wetland, a new namelist entry has been added:
wetland_option = 1 : for single point
wetland_option = 2 : for 2D regional

Initial test with the refactor wetland code at single point is conducted with the Fen site data from SK, Canada, for three years (see attached pdf).

[Test_Wetland_Code_Output.pdf](https://github.com/NCAR/noahmp/files/12433713/Test_Wetland_Code_Output.pdf)

CharlesZheZhang commented:
Thanks for updating this PR from my previous one. I approve this pull request.

cenlinhe commented:
point scale 1-year test simulations are successful on Derecho.

cenlinhe commented:
@tslin2 @RAbolafiaRosenzweig Could either of you approve PR if everything looks good? Thanks!

tslin2 commented on drivers/hrldas/WaterVarInTransferMod.F90:
```diff
@@ -205,6 +211,12 @@ subroutine WaterVarInTransfer(noahmp, NoahmpIO)
 
+    ! spatial varying wetland parameters from input
+    if ( noahmp%config%nmlist%OptWetlandModel == 2 ) then
+       noahmp%water%param%SoilSfcSatFracMax      = NoahmpIO%FSATMX(I,J)
```
Does this parameter need to be set when OptWetlandModel == 1 ?

tslin2 commented on parameters/NoahmpTable.TBL:
```diff
@@ -476,6 +476,8 @@
  Z0SOIL           = 0.002       ! Bare-soil roughness length (m) (i.e., under the canopy)
  Z0LAKE           = 0.01        ! Lake surface roughness length (m)
+ ! wetland parameter
+ WCAP             = 0.10        ! maximum wetland water holding capacity [m] (tunable) from Zhang et al. 2022
```
FSATMX           = 0.38        ! maximum surface saturated fraction (global mean)

tslin2 replied:
no worries. I saw it was there

tslin2 commented on drivers/lis/NoahmpIOVarType.F90:
```diff
@@ -345,6 +347,13 @@ module NoahmpIOVarType
     integer                                                ::  STEPWTD             ! step of groundwater call
 
+!------------------------------------------------------------------------
+! Needed for wetland model (OPT_WETLAND=1)
```
this should change to (the same as hrldas one): 
Needed for wetland model (OPT_WETLAND=1 or 2)

cenlinhe replied:
yes, you are right. I will correct it.

tslin2 commented on drivers/lis/WaterVarInTransferMod.F90:
```diff
@@ -154,6 +159,7 @@ subroutine WaterVarInTransfer(noahmp, NoahmpIO, LISparam)
     noahmp%water%param%DrainWatDepToImperv                = LISparam%TD_D
     noahmp%water%param%NumSoilLayerRoot                   = LISparam%NROOT
     noahmp%water%param%SoilDrainSlope                     = LISparam%SLOPE
+    noahmp%water%param%WetlandCapMax                      = LISparam%WCAP
```
might need to add WCRP in drivers/lis/LisNoahmpParamType.F90

cenlinhe replied:
thank you for catching this. You are right. I will correct it.

tslin2 commented on src/SoilWaterMainMod.F90:
```diff
@@ -148,6 +150,7 @@ subroutine SoilWaterMain(noahmp)
     if ( OptRunoffSurface == 6 ) call RunoffSurfaceVIC(noahmp,SoilTimeStep)
     if ( OptRunoffSurface == 7 ) call RunoffSurfaceXinAnJiang(noahmp,SoilTimeStep)
     if ( OptRunoffSurface == 8 ) call RunoffSurfaceDynamicVic(noahmp,SoilTimeStep,InfilSfcAcc)
+    if ( OptWetlandModel  >  0 ) call RunoffSurfaceWetland(noahmp)
```
@CharlesZheZhang @cenlinhe The wetland serves as a substitute for surface runoff and infiltration from other runoff schemes. Given that wetland runoff does not accumulate from other surface runoff schemes, should this consider wetlands as a separate runoff scheme? 

In addition, should wetland be default activated in wetland only land cover types, and in other land cover types use different runoff scheme? 
like usgs  
17: Herbaceous Wetland
18: Wooded Wetland

and modis
11: Permanent wetlands

cenlinhe replied:
This wetland runoff scheme is essentially the same as option 1 but with the FSAT as a input map or parameter directly instead of being parametrized in option 1. So I think it is ok to not adding a new runoff option. But you are right that it is better to have a land type if-statement to only allow wetland treatment over those wetland type grids. I am going to add this capability.

CharlesZheZhang replied:
The wetland flag should not based on land use type - since this scheme is to simulate subgrid wetland (potholes), so the dominant land use may not be wetland, but could be cropland and forest, etc. 
The wetland flag on 2D model run should base on read in variable FSAT and WCAP - meaning if the maximum saturated fraction > 0 and maximum storage > 0, it will generate wetland ponding water from runoff, etc.

cenlinhe replied:
OK, then I am going to remove the wetland flag control on the wetland module but I will keep the wetland flag variable in the model just in case it will be used in the future.

cenlinhe replied:
Since it is a subgrid wetland treatment, I am going to remove the wetland flag control on those modules for now (but keep the wetland flag variable for future use). In the future, we may need to add a new input map to identify those grids where subgrid wetlands exist and give them the wetland flag. Currently, it is difficult to have wetland flag defined based on dynamically changing FSAT and WCAP. If this looks good to both of you, I am going to make this change.

tslin2 replied:
sounds good

tslin2 replied:
Should these two maps of parameters be added to somewhere's directories for downloading if data is available?

cenlinhe replied:
Yes, it will be added to a unified Noah-MP database in the future, but I do not think we have a global map for these two parameters ready to use.

tslin2 commented on src/WetlandWaterZhang22Mod.F90:
```diff
@@ -0,0 +1,102 @@
+    ! compute wetland water balance
+    LatHeatPot = 1.26 * VapPresSlope * (RadSwAbsSfc - RadLwNetSfc - HeatGroundTot) / &
+                 (VapPresSlope + PsychroConst)                                           ! LatHeatPot Potential latent heat W/m2 P-T method
+    WaterStorageWetland = WaterStorageWetland + RunoffSurface     
```
@CharlesZheZhang @cenlinhe should soil moisture be updated to be saturated or wet?

tslin2 replied:
I guess my question is, is this wetland process considers as a subgrid process?

cenlinhe replied:
I think the soil moisture does not have to be fully saturated or to be updated here. It is still dependent on the regular Noah-MP process. This simplified wetland scheme basically adds a ponding/water storage for the saturated subgrid portion. It is mainly controlled by the saturation fraction if I understand correctly. I will let @CharlesZheZhang answer this question.

cenlinhe replied:
It is not a comprehensive wetland scheme, which does not involve any explicit treatment of soil below wetland ponding water and will need future improvement.

tslin2 replied:
got it, thanks!

cenlinhe commented:
@CharlesZheZhang @tslin2 I have updated the code to include a wetland flag and fixed all the issues mentioned in your comments. Could you please take a look and see if it can be approved?

cenlinhe commented:
OK, I have fixed all the issues mentioned above, please take a look to see if it can be approved. thanks!

cenlinhe commented on src/RunoffSurfaceWetlandMod.F90:
```diff
@@ -0,0 +1,50 @@
+! ----------------------------------------------------------------------
+
+    ! compute saturated area fraction, based on top-layer soil moisture
+    SoilSaturateFrac = SoilSfcSatFracMax * (SoilLiqWater(1)/SoilMoistureSat(1))
```
Yes, currently they will use the same value. I agree that we probably need a separate parameter name with a regional-calibrated value for wetland. This can be done in the future.

tslin2 replied:
I think no matter what, this wetland scheme would replace surface runoff and infiltration on top of other surface runoff schemes; it may be better to consider this as another surface runoff scheme given the runoff schemes will be useless

cenlinhe replied:
agreed. This issue will be solved once we have a wetland flag to identify the land type to only do wetland over grids with existing subgrid wetland areas.

tslin2 replied:
Or maybe add the if statement only if SoilSfcSatFracMax >0 to run the wetland module to prevent running wetland for everywhere

cenlinhe replied:
This is one possible way but only works for 2D FSATMX input map not for using a uniform value from NoahmpTable.TBL.
